src/com/mendix/widget/carousel/components/Carousel.ts

```diff
@@ -11,16 +13,65 @@ export interface CarouselProps {
     images?: Image[];
 }
 
-const carouselItems = (images: Image[] = []) =>
-    images.map((image, index) => createElement(CarouselItem, {
-        active: index === 0,
-        key: index,
-        url: image.url
-    }));
-
-export const Carousel = (props: CarouselProps) =>
-    DOM.div({ className: "widget-carousel" },
-        DOM.div({ className: "widget-carousel-item-wrapper" },
-            carouselItems(props.images)
-        )
-    );
+interface CarouselState {
+    activeIndex: number;
+}
+
+type Direction = "right" | "left";
+
+export class Carousel extends Component<CarouselProps, CarouselState> {
+    static defaultProps: CarouselProps = { images: [] };
+
+    constructor(props: CarouselProps) {
+        super(props);
+
+        this.state = { activeIndex: 0 };
+    }
+
+    render() {
+        return DOM.div({ className: "widget-carousel" },
+            DOM.div({ className: "widget-carousel-item-wrapper" },
+                this.getCarouselItems(this.props.images, this.state.activeIndex)
+            ),
+            this.props.images.length ? this.getCarouselControls() : null
+        );
+    }
+
+    private getCarouselItems(images: Image[], activeIndex: number) {
+        return images.map((image, index) =>
+            createElement(CarouselItem, {
+                active: index === activeIndex,
+                key: index,
+                url: image.url
+            }));
+    }
+
+    private getCarouselControls() {
+        return [
+            createElement(CarouselControl, {
+                direction: "left",
+                key: 0,
+                onClick: () => this.moveInDirection("left")
```

Comment:
This actually allocates memory for and creates a new closure every time `render` is called. That is not needed, just create it once in the constructor and reuse it.